### PR TITLE
basedpyright: 1.17.5 -> 1.18.0

### PR DESCRIPTION
**Pull request by kiike**
## Description of changes

The package now needs a new dependency `@detachhead/ts-helpers`, so the `package.json` and the lockfile must be changed accordingly.

I was thinking of creating the `package-lock.json` in another derivation. Would that be OK?

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by kiike** [comment]
```
 cat << EOF > test.py; ./result/bin/basedpyright --version; ./result/bin/basedpyright test.py
import sys

if sys.platform == "win32":
    a = "a" + 1

print(3)
noprint(3)

EOF
basedpyright 1.18.0
based on pyright 1.1.382
/home/kiike/projects/nixpkgs/test.py
  /home/kiike/projects/nixpkgs/test.py:4:5 - error: Type of "a" is unknown (reportUnknownVariableType)
  /home/kiike/projects/nixpkgs/test.py:4:9 - error: Operator "+" not supported for types "Literal['a']" and "Literal[1]" (reportOperatorIssue)
  /home/kiike/projects/nixpkgs/test.py:7:1 - error: "noprint" is not defined (reportUndefinedVariable)
3 errors, 0 warnings, 0 notes
```

**Comment by kiike** [comment]
@GaetanLepage

**Comment by kiike** [comment]
```
cat << EOF > test.py; ./result/bin/basedpyright --version; ./result/bin/basedpyright test.py
import sys

if sys.platform == "win32":
    a = "a" + 1

print(3)
noprint(3)

EOF
basedpyright 1.18.0
based on pyright 1.1.382
/home/kiike/projects/nixpkgs/test.py
  /home/kiike/projects/nixpkgs/test.py:4:5 - error: Type of "a" is unknown (reportUnknownVariableType)
  /home/kiike/projects/nixpkgs/test.py:4:9 - error: Operator "+" not supported for types "Literal['a']" and "Literal[1]" (reportOperatorIssue)
  /home/kiike/projects/nixpkgs/test.py:7:1 - error: "noprint" is not defined (reportUndefinedVariable)
3 errors, 0 warnings, 0 notes
```

File: pkgs/by-name/ba/basedpyright/package.nix [hunk 1]
```diff
@@ -11,26 +11,26 @@
...
-    hash = "sha256-DaNxPGY0ahTcluCnsEZNL2oP9zKWQlON8i1bxeJ5GnU=";
+    hash = "sha256-o2MHZMUuVnVjdv2b+GLIMjK1FT8KfLUzo7+zH7OU7HI=";
   };
 
   patchedPackageJSON = runCommand "package.json" { } ''
     ${jq}/bin/jq '
-      .devDependencies |= with_entries(select(.key == "glob" or .key == "jsonc-parser"))
+      .devDependencies |= with_entries(select(.key == "glob" or .key == "jsonc-parser" or .key == "@detachhead/ts-helpers"))
       | .scripts =  {  }
       ' ${src}/package.json > $out
```
Comment (kiike) [inline pkgs/by-name/ba/basedpyright/package.nix:27]: I'd like to do something like 

```nix
${nodejs}/bin/npm update --package-lock
```

**Review by kiike — COMMENTED** [review]
Let me know what you think of the suggestion.

**Comment by GaetanLepage** [comment]
> The package now needs a new dependency `@detachhead/ts-helpers`, so the `package.json` and the lockfile must be changed accordingly.
> 
> I was thinking of creating the `package-lock.json` in another derivation. Would that be OK?

I am unfortunately not familiar at all with the node ecosystem. Maybe, try to ask someone in the _nixpkgs contributors_ room.
But anyway, if you are confident in your choice and if it works fine, let's go for it, we don't need to ask the permission to anyone else.

**Comment by GaetanLepage** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 344536`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>basedpyright</li>
    <li>emacsPackages.lsp-bridge</li>
  </ul>
</details>

---
### `aarch64-linux`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>basedpyright</li>
    <li>emacsPackages.lsp-bridge</li>
  </ul>
</details>

---
### `x86_64-darwin`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>basedpyright</li>
    <li>emacsPackages.lsp-bridge</li>
  </ul>
</details>

---
### `aarch64-darwin`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>basedpyright</li>
    <li>emacsPackages.lsp-bridge</li>
  </ul>
</details>

**Comment by kiike** [comment]
I think the approach I took of commiting the lockfile is what is used in nixpkgs. Updating the lockfile inside the derivation I think could result in reproducibility errors, and adding the lockfile is the way to go for now. What I don't like is that the lockfile is gigantic and hard to audit.

I added a comment on `package.nix` to indicate how to update the `package-lock.json` when needed. The package should be good to go now.

**Comment by GaetanLepage** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 344536`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>basedpyright</li>
    <li>emacsPackages.lsp-bridge</li>
  </ul>
</details>

---
### `aarch64-linux`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>basedpyright</li>
    <li>emacsPackages.lsp-bridge</li>
  </ul>
</details>

---
### `x86_64-darwin`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>basedpyright</li>
    <li>emacsPackages.lsp-bridge</li>
  </ul>
</details>

---
### `aarch64-darwin`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>basedpyright</li>
    <li>emacsPackages.lsp-bridge</li>
  </ul>
</details>

**Review by GaetanLepage — APPROVED** [review]
LGTM !

**Pull request merged**